sources/parsers/src/Evaluator.cpp

```diff
@@ -0,0 +1,202 @@
+/// to deal in the Software without restriction, including without limitation
+/// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+/// and/or sell copies of the Software, and to permit persons to whom the
+/// Software is furnished to do so, subject to the following conditions:
+///
+/// The above copyright notice and this permission notice shall be included
+/// in all copies or substantial portions of the Software.
+///
+/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+/// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+/// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+/// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+/// IN THE SOFTWARE.
+
+#include <sstream>
+
+#include "Evaluator.h"
+
+#undef yyFlexLexer
+/// Decimal lexer definition.
+#define yyFlexLexer DecimalFlexLexer
+#include <FlexLexer.h>
+#include "DecimalScanner.h"
+
+#undef yyFlexLexer
+/// Float lexer definition.
+#define yyFlexLexer FloatFlexLexer
+#include <FlexLexer.h>
+#include "FloatScanner.h"
+
+#undef yyFlexLexer
+/// String lexer definition.
+#define yyFlexLexer StringFlexLexer
+#include <FlexLexer.h>
+#include "StringScanner.h"
+
+PARSER::Evaluator::Evaluator(const std::string &rSourceName) :
+    m_rSourceName(rSourceName),
+    m_Context(nullptr, 0)
+{
+}
+
+void PARSER::Evaluator::ParseDecimalString(const std::string &input)
+{
+    std::istringstream iss(input);
+    if (!ParseDecimalStream(iss))
+    {
+        throw PARSER::EXCEPTIONS::ParsingError(ParsingErrorMessage(input));
+    }
+    m_Expression = input;
+}
+
+void PARSER::Evaluator::ParseFloatString(const std::string &input)
+{
+    std::istringstream iss(input);
+    if (!ParseFloatStream(iss))
+    {
+        throw PARSER::EXCEPTIONS::ParsingError(ParsingErrorMessage(input));
+    }
+    m_Expression = input;
+}
+
+void PARSER::Evaluator::ParseStringString(const std::string &input)
+{
+    std::istringstream iss(input);
+    if (!ParseStringStream(iss))
+    {
+        throw PARSER::EXCEPTIONS::ParsingError(ParsingErrorMessage(input));
+    }
+    m_Expression = input;
+}
+
+void PARSER::Evaluator::SetContext(const std::smatch *pGroups, uint64_t lineNo) const
+{
+    m_Context.Set(pGroups, lineNo);
+}
+
+PARSER::ExpressionContext &PARSER::Evaluator::GetContext() const
+{
+    return m_Context;
+}
+
+std::tuple<double, std::string> PARSER::Evaluator::EvaluateDouble() const
+{
+    double value;
+    std::string stringValue;
+    try
+    {
+        std::tie(value, stringValue) = m_Context.GetExpression()->EvaluateDouble();
+    }
+    catch (const EXCEPTIONS::EvaluatorException &evaluatorError)
+    {
+        throw PARSER::EXCEPTIONS::EvaluatorException(EvaluationErrorMessage(evaluatorError.what()));
```

Comment:
Catch only to rethrow in weird way?
(Probably in many other places that I didn't notice)

Reply:
Well... Why can't receiving end use `EvaluationErrorMessage` when needed?

Reply:
Ok, I understand.